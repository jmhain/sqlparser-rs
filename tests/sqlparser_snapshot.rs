// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#![warn(clippy::all)]

use sqlparser::ast::Statement;
use sqlparser::dialect::{Dialect, GenericDialect};
use sqlparser::parser::Parser;
use std::path::Path;

fn test_dialect(path: &Path, dialect: &dyn Dialect) -> datatest_stable::Result<()> {
    let sql = std::fs::read_to_string(path)?;

    let stmts = Parser::parse_sql(dialect, &sql)?;
    let name = path
        .strip_prefix(Path::new("tests/queries"))?
        .to_str()
        .unwrap();
    insta::assert_yaml_snapshot!(name, stmts);

    // Rendering the AST to a string and re-parsing it should yield the same AST.
    for stmt in stmts {
        let [round_trip_stmt]: [Statement; 1] = Parser::parse_sql(dialect, &stmt.to_string())?
            .try_into()
            .map_err(|_| "expected a single statement")?;
        pretty_assertions::assert_eq!(stmt, round_trip_stmt);
    }

    Ok(())
}

fn test_generic(path: &Path) -> datatest_stable::Result<()> {
    test_dialect(path, &GenericDialect)
}

datatest_stable::harness!(test_generic, "tests/queries/generic", r"^.*.sql",);
