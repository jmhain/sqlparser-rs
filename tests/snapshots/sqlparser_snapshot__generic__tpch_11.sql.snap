---
source: tests/sqlparser_snapshot.rs
expression: stmts
---
- Query:
    with: ~
    body:
      Select:
        distinct: ~
        top: ~
        projection:
          - UnnamedExpr:
              Identifier:
                value: ps_partkey
                quote_style: ~
          - ExprWithAlias:
              expr:
                Function:
                  name:
                    - value: sum
                      quote_style: ~
                  args:
                    - Unnamed:
                        Expr:
                          BinaryOp:
                            left:
                              Identifier:
                                value: ps_supplycost
                                quote_style: ~
                            op: Multiply
                            right:
                              Identifier:
                                value: ps_availqty
                                quote_style: ~
                  filter: ~
                  null_treatment: ~
                  over: ~
                  distinct: false
                  special: false
                  order_by: []
              alias:
                value: value
                quote_style: ~
        into: ~
        from:
          - relation:
              Table:
                name:
                  - value: partsupp
                    quote_style: ~
                alias: ~
                args: ~
                with_hints: []
                version: ~
                partitions: []
            joins: []
          - relation:
              Table:
                name:
                  - value: supplier
                    quote_style: ~
                alias: ~
                args: ~
                with_hints: []
                version: ~
                partitions: []
            joins: []
          - relation:
              Table:
                name:
                  - value: nation
                    quote_style: ~
                alias: ~
                args: ~
                with_hints: []
                version: ~
                partitions: []
            joins: []
        lateral_views: []
        selection:
          BinaryOp:
            left:
              BinaryOp:
                left:
                  BinaryOp:
                    left:
                      Identifier:
                        value: ps_suppkey
                        quote_style: ~
                    op: Eq
                    right:
                      Identifier:
                        value: s_suppkey
                        quote_style: ~
                op: And
                right:
                  BinaryOp:
                    left:
                      Identifier:
                        value: s_nationkey
                        quote_style: ~
                    op: Eq
                    right:
                      Identifier:
                        value: n_nationkey
                        quote_style: ~
            op: And
            right:
              BinaryOp:
                left:
                  Identifier:
                    value: n_name
                    quote_style: ~
                op: Eq
                right:
                  Value:
                    SingleQuotedString: GERMANY
        group_by:
          Expressions:
            - Identifier:
                value: ps_partkey
                quote_style: ~
        cluster_by: []
        distribute_by: []
        sort_by: []
        having:
          BinaryOp:
            left:
              Function:
                name:
                  - value: sum
                    quote_style: ~
                args:
                  - Unnamed:
                      Expr:
                        BinaryOp:
                          left:
                            Identifier:
                              value: ps_supplycost
                              quote_style: ~
                          op: Multiply
                          right:
                            Identifier:
                              value: ps_availqty
                              quote_style: ~
                filter: ~
                null_treatment: ~
                over: ~
                distinct: false
                special: false
                order_by: []
            op: Gt
            right:
              Subquery:
                with: ~
                body:
                  Select:
                    distinct: ~
                    top: ~
                    projection:
                      - UnnamedExpr:
                          BinaryOp:
                            left:
                              Function:
                                name:
                                  - value: sum
                                    quote_style: ~
                                args:
                                  - Unnamed:
                                      Expr:
                                        BinaryOp:
                                          left:
                                            Identifier:
                                              value: ps_supplycost
                                              quote_style: ~
                                          op: Multiply
                                          right:
                                            Identifier:
                                              value: ps_availqty
                                              quote_style: ~
                                filter: ~
                                null_treatment: ~
                                over: ~
                                distinct: false
                                special: false
                                order_by: []
                            op: Multiply
                            right:
                              Value:
                                Number:
                                  - "0.0001000000"
                                  - false
                    into: ~
                    from:
                      - relation:
                          Table:
                            name:
                              - value: partsupp
                                quote_style: ~
                            alias: ~
                            args: ~
                            with_hints: []
                            version: ~
                            partitions: []
                        joins: []
                      - relation:
                          Table:
                            name:
                              - value: supplier
                                quote_style: ~
                            alias: ~
                            args: ~
                            with_hints: []
                            version: ~
                            partitions: []
                        joins: []
                      - relation:
                          Table:
                            name:
                              - value: nation
                                quote_style: ~
                            alias: ~
                            args: ~
                            with_hints: []
                            version: ~
                            partitions: []
                        joins: []
                    lateral_views: []
                    selection:
                      BinaryOp:
                        left:
                          BinaryOp:
                            left:
                              BinaryOp:
                                left:
                                  Identifier:
                                    value: ps_suppkey
                                    quote_style: ~
                                op: Eq
                                right:
                                  Identifier:
                                    value: s_suppkey
                                    quote_style: ~
                            op: And
                            right:
                              BinaryOp:
                                left:
                                  Identifier:
                                    value: s_nationkey
                                    quote_style: ~
                                op: Eq
                                right:
                                  Identifier:
                                    value: n_nationkey
                                    quote_style: ~
                        op: And
                        right:
                          BinaryOp:
                            left:
                              Identifier:
                                value: n_name
                                quote_style: ~
                            op: Eq
                            right:
                              Value:
                                SingleQuotedString: GERMANY
                    group_by:
                      Expressions: []
                    cluster_by: []
                    distribute_by: []
                    sort_by: []
                    having: ~
                    named_window: []
                    qualify: ~
                order_by: []
                limit: ~
                limit_by: []
                offset: ~
                fetch: ~
                locks: []
        named_window: []
        qualify: ~
    order_by:
      - expr:
          Identifier:
            value: value
            quote_style: ~
        asc: false
        nulls_first: ~
    limit: ~
    limit_by: []
    offset: ~
    fetch: ~
    locks: []

